[Flang] Maxloc elemental intrinsic lowering.

davemgreen opened the pull request:
This is an extension to #74828 to handle maxloc too, to keep the minloc and maxloc symmetric.

tblah reviewed (approved):
Looks good to me. Thank you for adding all of these!

tblah commented on flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:
```diff
@@ -812,50 +812,55 @@ class ReductionElementalConversion : public mlir::OpRewritePattern<Op> {
-                 rank, hlfir::getFortranElementType(minloc.getType())));
+                 rank, hlfir::getFortranElementType(mloc.getType())));
 
     auto init = [](fir::FirOpBuilder builder, mlir::Location loc,
```
Windows CI needs isMax to be added to the capture specification

davemgreen committed:
[Flang] Maxloc elemental intrinsic lowering.

This is an extension to #74828 to handle maxloc too, to keep the minloc and
maxloc symmetric.

davemgreen commented:
Thanks. I'm surprised this didn't come up in both compilers, I hadn't realized it would treat constexpr differently in lambda expressions.

yi-wu-arm commented:
The [flang-aarch64-libcxx](https://lab.llvm.org/buildbot/#/builders/268)[7035](https://lab.llvm.org/buildbot/#/builders/268/builds/7035) build bot is failing: https://lab.llvm.org/buildbot/#/builders/268/builds/7035, but I'm not sure if that is related to this patch.

davemgreen commented:
Yep that is me. I was just about to push a patch so long as it passed testing locally.

The MSVC compiler didn't like things one way, and the -Weverything build didn't like it the other :(
